packages/@react-aria/focus/src/FocusScope.tsx

```diff
@@ -119,30 +119,35 @@ export function useFocusManager(): FocusManager {
 function createFocusManager(scopeRef: React.RefObject<HTMLElement[]>): FocusManager {
   return {
     focusNext(opts: FocusManagerOptions = {}) {
-      let node = opts.from || document.activeElement;
-      let focusable = getFocusableElementsInScope(scopeRef.current, opts);
-      let nextNode = focusable.find(n =>
-        !!(node.compareDocumentPosition(n) & (Node.DOCUMENT_POSITION_FOLLOWING | Node.DOCUMENT_POSITION_CONTAINED_BY))
-      );
-      if (!nextNode && opts.wrap) {
-        nextNode = focusable[0];
+      let scope = scopeRef.current;
+      let {from, tabbable, wrap} = opts; 
+      let node = from || document.activeElement;
+      let sentinel = scope[0].previousElementSibling;
+      let walker = getFocusableTreeWalker(getScopeRoot(scope), {tabbable});
+      walker.currentNode = isElementInScope(node, scope) ? node : sentinel;
+      let nextNode = walker.nextNode() as HTMLElement;
+      if (!(nextNode && isElementInScope(nextNode, scope)) && wrap) {
+        walker.currentNode = sentinel;
+        nextNode = walker.nextNode() as HTMLElement;
       }
-      if (nextNode) {
-        nextNode.focus();
+      if (nextNode && isElementInScope(nextNode, scope)) {
+        focusElement(nextNode, true);
       }
       return nextNode;
     },
     focusPrevious(opts: FocusManagerOptions = {}) {
-      let node = opts.from || document.activeElement;
-      let focusable = getFocusableElementsInScope(scopeRef.current, opts).reverse();
+      let scope = scopeRef.current;
+      let {from, tabbable, wrap} = opts; 
+      let node = from || document.activeElement;
+      let focusable = getFocusableElementsInScope(scope, {tabbable}).reverse();
       let previousNode = focusable.find(n =>
         !!(node.compareDocumentPosition(n) & (Node.DOCUMENT_POSITION_PRECEDING | Node.DOCUMENT_POSITION_CONTAINED_BY))
       );
-      if (!previousNode && opts.wrap) {
+      if (!previousNode && wrap) {
         previousNode = focusable[0];
       }
       if (previousNode) {
-        previousNode.focus();
+        focusElement(previousNode, true);
       }
       return previousNode;
     }
```

Comment:
TODO refactor to use tree walking starting from end sentinel, while still accounting for `!!(node.compareDocumentPosition(n) & (Node.DOCUMENT_POSITION_PRECEDING | Node.DOCUMENT_POSITION_CONTAINED_BY))` filter.

Reply:
I have refactored this with cce1045a1f2125173f042c231e9f58cbfc722509.